asf_online_data_exploration/pipeline/data_collection/the_guardian.py

```diff
@@ -0,0 +1,151 @@
+    response_status_code = response.status_code
+    if response_status_code != 200:
+        if response_status_code >= 400 and response_status_code < 500:
+            raise Exception(
+                "Cannot get data, the program will stop!\nHTTP {}: {}".format(
+                    response_status_code, response.text
+                )
+            )
+
+        sleep_seconds = random.randint(5, 60)
+        print(
+            "Cannot get data, your program will sleep for {} seconds...\nHTTP {}: {}".format(
+                sleep_seconds, response_status_code, response.text
+            )
+        )
+        time.sleep(sleep_seconds)
+        return connect_to_endpoint(endpoint_url)
+    # sleep 1 second not to surpass rate limit
+    time.sleep(1)
+    return response.json()["response"]
+
+
+def collect_and_process_guardian_data(
+    api_key, rules, query_params, s3_bucket: str, s3_folder: str
+):
+    """
+    Collects, processes and saves The Guardian data following a set of rules and query parameters.
+
+    Args:
+        api_key: API key credentials
+        rules: rules for collecting Twitter data. Each rules should contain "value" and "tag" keys
+        query_paramers: parameters for data collection
+    """
+    if ("from-date" in query_params.keys()) and ("from-date" in query_params.keys()):
+        from_date = query_params["from-date"].replace("/", "_")
+        to_date = query_params["to-date"].replace("/", "_")
+        specified_time_frame_flag = True
+    else:
+        specified_time_frame_flag = False
+
+    # data collection for every possible rule
+    for i in range(len(rules)):
+        query_params["q"] = rules[i]["value"]
+        query_tag = rules[i]["tag"]
+
+        endpoint_url = define_endpoint_url(api_key, query_params)
+
+        # Collecting and processing data
+        json_response = connect_to_endpoint(endpoint_url)
+        data = json_response["results"]
+
+        # Extracting parameters from json response
+        total_pages = json_response["pages"]
+        current_page = json_response["currentPage"]  # this should be 1
+
+        # Define a filename and uploading data to s3
+        if specified_time_frame_flag:
+            filename = f"guardian_{query_tag}_S{from_date}F{to_date}.json"
+        else:
+            filename = f"guardian_{query_tag}_{datetime.now()}.json"
+        dictionary_to_s3(data, s3_bucket, s3_folder, filename)
+
+        current_page += 1
+        while current_page <= total_pages:
+            # Adding the current page to the query parameters
+            query_params["page"] = current_page
+            endpoint_url = define_endpoint_url(api_key, query_params)
+
+            # Collecting and processing data
+            json_response = connect_to_endpoint(endpoint_url)
+            data = data + json_response["results"]
+
+            # Uploading to S3
+            dictionary_to_s3(data, s3_bucket, s3_folder, filename)
+
+            current_page += 1
```

Comment:
The only reason for saving at every iteration was just so that no data would be lost in case there's a problem with the program while running, since there is a limit number of calls to the API (and hence, not having to collect the same data twice). But yeah, I'll change it, it makes the program much slower! :D

Reply:
ah yes that makes sense!